expression/constant_propagation.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// MaxPropagateColsCnt means the max number of columns that can participate propagation.
+var MaxPropagateColsCnt = 100
+
+var eqFuncNameMap = map[string]bool{
+	ast.EQ: true,
+}
+
+// inEqFuncNameMap stores all the in-equal operators that can be propagated.
+var inEqFuncNameMap = map[string]bool{
+	ast.LT: true,
+	ast.GT: true,
+	ast.LE: true,
+	ast.GE: true,
+	ast.NE: true,
+}
+
+type propagateConstantSolver struct {
+	colMapper        map[string]int // colMapper maps column to its index
+	transitiveMatrix [][]bool       // transitiveMatrix[i][j] = true means we can infer that col i = col j
+	eqList           []*Constant    // if eqList[i] != nil, it means col i = eqList[i]
+	columns          []*Column      // columns stores all columns appearing in the conditions
+	conditions       []Expression
+	foreverFalse     bool
+}
+
+// propagateInEQ propagates all in-equal conditions.
+// e.g. For expression a = b and b = c and c = d and c < 1 , we can get a < 1 and b < 1.
+// We maintain a matrix representing the equivalent for every two columns.
+func (s *propagateConstantSolver) propagateInEQ() {
```

Comment:
Why use floyd to construct the transitiveMatrix intead of union-find?
time complexity of constructing the matrix seems be higher than constructing a union-find structure?